### PR TITLE
Support gradient clipping

### PR DESCRIPTION
**Pull request by dding3**
## What changes were proposed in this pull request?
1. support clip by value
2. support l2 norm clip

## How was this patch tested?
manual test

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala [hunk 1]
```diff
@@ -292,16 +292,17 @@ class AllReduceParameter[T: ClassTag](id: Long, partitionNum: Int, size: Int) ex
    */
   def sendWeightPartition(): Unit = {
     val blockId = getWeightBlockId(partitionId)
+    val localBuffer = BlockManagerWrapper.getLocalBytes(blockId).getOrElse {
+      throw new RuntimeException(s"Didn't find weight block $blockId in the block " +
+        s"manager. Did you initialize this AllReduceParameter on every executor?")
+    }
+    SerializerInstance.serialize(localBuffer).compress(weightPartition)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: The `serialize(ByteBuffer)` method is confusing; maybe change to a different name such as create

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala [hunk 2]
```diff
@@ -152,8 +152,19 @@ class LocalOptimizer[T: ClassTag] (
           })
       )
       val loss = lossSum / parallelism
-      grad.div(ev.fromType(parallelism))
+      var scale = ev.fromType(parallelism)
+      if (gradientClippingParams.enableL2NormClipping) {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala]: If the model (e.g., AlexNet or VGG) has large parameter sizes, the overhead can be large? Shall we do this using multi-threading?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 3]
```diff
@@ -289,20 +297,43 @@ object DistriOptimizer {
         numFinishedModelUpdates >= driverSubModelNum * (1.0 - maxDropPercentage)) {
         // enough records were processed for this batch, so update the model
         val value = lossSum.value / numFinishedModelUpdates
+
+        var l2Norm = 0.0f
+        var scale = ev.fromType(numFinishedModelUpdates)
+        if (normClippingEnable) {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: If the model (e.g., AlexNet or VGG) has large parameter sizes, the overhead can be large? Shall we do this using multi-threading?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 4]
```diff
@@ -452,3 +481,10 @@ object Optimizer {
     }
   }
 }
+
+case class GradientClippingParams(
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: make this a trait, and implement two subclass for l2norm and constant clipping respectively

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala [hunk 5]
```diff
@@ -866,6 +866,34 @@ private[tensor] class DenseTensor[@specialized T: ClassTag](
...
   }
 
+  protected def sumNSquare(n: T): T = {
+    var res: T = ev.fromType(0)
+    val func = new TensorFunc2[T] {
+      override def apply(data1: Array[T], offset1: Int): Unit = {
+        res = ev.plus(res, ev.pow(ev.abs(data1(offset1)), n))
+      }
+    }
+    DenseTensorApply.apply1[T](this, func)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: Instead of using apply1, how about using sth like vsMul?

**Comment by jason-dai** [comment]
Please verify the performance overhead on models with large parameter sizes (such as AlexNet or LSTM)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 6]
```diff
@@ -148,6 +149,13 @@ object DistriOptimizer {
     var dropModelNumBatch = 0
     var lossArray = new Array[Double](_subModelNumber)
 
+    // gradient clip settings
+    val constantClippingEnable = clippingParams.enableConstantClipping
+    val normClippingEnable = clippingParams.enableL2NormClipping
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: what if both l2norm and constant clipping are set to true?

Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: It will use l2norm first and then constant clipping

Comment (jason-dai) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Does it make sense?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala (hunk 2 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala]: Updated to use multi-thread to compute the square sum

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 3 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Updated

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala (hunk 5 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: Updated to use MKL

File: pyspark/bigdl/optim/optimizer.py [hunk 7]
```diff
@@ -669,6 +669,25 @@ def set_checkpoint(self, checkpoint_trigger,
         callBigDlFunc(self.bigdl_type, "setCheckPoint", self.value,
                       checkpoint_trigger, checkpoint_path, isOverWrite)
 
+    def set_constant_clipping(self, min_value, max_value):
```
Comment (shane-huang) [inline pyspark/bigdl/optim/optimizer.py]: shall we change the name to gradient clipping? We can not know what it is clipping only by the method name. And add some docs?

Comment (shane-huang) [reply pyspark/bigdl/optim/optimizer.py]: if the name too long, u can use sth. like set_gradclip_const, and set_gradclip_l2norm

Comment (jason-dai) [reply pyspark/bigdl/optim/optimizer.py]: Sounds good

**Comment by shane-huang** [comment]
BTW, is it possible to make this pr more general to allow more customization of gradients handing? Actually I need a little bit of more processing besides clipping. Can I set a callback Tensor -> Tensor to define my own processing of gradients before they are passed to OptimMethod? Better be a trait though, so I will be able to set some member variables.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 6 above)
Comment (shane-huang) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: or shall we make it a single interface and the method of clipping as parameters? e.g. interface setGradClip(['const']) or setGradClip(['l2']) if I want one of the two, and if I want to stack those I shall write setGradClip(['const', 'l2']), the order also matters.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 8]
```diff
@@ -284,6 +286,33 @@ abstract class Optimizer[T: ClassTag, D](
...
 
   def prepareInput(): Unit = {}
+
+  /**
+   * Set constant gradient clipping
+   * @param min the minimum value to clip by
+   * @param max the maximum value to clip by
+   * @return
+   */
+  def setConstantGradientClipping(min: Float, max: Float)
```
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: any way to disable clipping? People might use clipping in some stages of training and then diable it later. Also optimizer could be reused. Need a way to set to defaults.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 9]
```diff
@@ -284,6 +286,33 @@ abstract class Optimizer[T: ClassTag, D](
...
+    gradientClippingParams.maxValueClip = max
+    this
+  }
+
+  /**
+   * Clip gradient to a maximum L2-norm
+   * @param clipNorm gradient L2-Norm threshold
+   * @return
+   */
+  def setGradientClippingByNorm(clipNorm: Float)
```
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: method name should include L2, because there're several kinds of norms (e.g. l1)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala (hunk 4 above)
Comment (shane-huang) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: agree

**Comment by shane-huang** [comment]
Please check my comments.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala [hunk 10]
```diff
@@ -866,6 +866,38 @@ private[tensor] class DenseTensor[@specialized T: ClassTag](
...
+  }
+
+  override def clamp(min: Float, max: Float): Tensor[T] = {
+    val maxT = ev.fromType[Float](max)
+    val minT = ev.fromType[Float](min)
+    val func = new TensorFunc2[T] {
+      override def apply(data1: Array[T], offset1: Int): Unit = {
+        if (ev.isGreater(data1(offset1), maxT)) data1(offset1) = maxT
+        else if (ev.isGreater(minT, data1(offset1))) data1(offset1) = minT
+      }
```
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]:
```scala
if (ev.isGreater(data1(offset1), maxT)) data1(offset1) = maxT
else if (ev.isGreater(minT, data1(offset1))) data1(offset1) = minT
```
could be 
```scala
ev.min(ev.max(data1(offset, minT)), maxT)
```

File: pyspark/bigdl/optim/optimizer.py (hunk 7 above)
Comment (dding3) [reply pyspark/bigdl/optim/optimizer.py]: Updated

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala (hunk 8 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: Added disable clipping method.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala (hunk 9 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: Updated

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala (hunk 10 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: The latter code is much cleaner than the previous one. But I think it may affect performance since it need compare twice every time

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala [hunk 11]
```diff
@@ -152,8 +152,31 @@ class LocalOptimizer[T: ClassTag] (
...
+        }
+        val l2Norm = (math.sqrt(sum) / parallelism).toFloat
+        
+        if (l2Norm > gradientClippingParams.normValueClip) {
+          scale = ev.fromType[Float]((l2Norm * parallelism) / gradientClippingParams.normValueClip)
+        }
+      }
+      grad.div(scale)
 
+      if (gradientClippingParams.enableConstantClipping) {
```
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala]: possible to make it a trait? so that people could define their own clipping.

Comment (jason-dai) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala]: I think we can leave it for future improvement.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/Parameter.scala [hunk 12]
```diff
@@ -58,14 +58,14 @@ object SerializerInstance {
     this.pm = pm
   }
 
-  def serialize[T: ClassTag](data: Tensor[T]): CompressedTensor[T] = {
+  def create[T: ClassTag](data: Tensor[T]): CompressedTensor[T] = {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/Parameter.scala]: this is actually to serialize the tensor, not create

**Comment by jason-dai** [comment]
LGTM

**Comment by dding3** [comment]
Thanks for your review. Jekins passed.
